### PR TITLE
Baseline Correction

### LINKED ISSUE #296: Baseline Correction for Kinematic Variables
## Reason
Integration of a harmonic, or quasi-harmonic, acceleration or velocity time histories can result in displacements that exhibit an unrealistic deviation from a particle's initial position over time, depending on the assumed properties of the given signal. Initial conditions are often unknown or difficult to satisfy while maintaining the expected behavior that a vibration signal oscillates about a constant baseline and comes to rest at or near the initial position. Baseline correction is a method to transform the time history variables of a vibrating particle such that its mean position is a constant, zero-valued one, without adversely affecting the peak accelerations or spectral properties. I think it could be useful to many to have a C++ object capable of performing these corrections in MASTODON.

Implementation of this capability would be in support of my graduate research for [NEUP Project: 17-12939](https://neup.inl.gov/Lists/Awards/DispForm.aspx?ID=593&ContentTypeId=0x01000FF41AD33234E04B888509DF4B53DDA9).

## Design
The object would be a vectorpostprocessor whose general purpose would fall into the same category as the ResponseSpectraCalculator, i.e., a signal processing function. A user could pass a nodal acceleration solution and the vpp would return the adjusted time histories. They could also run the vpp on some input acceleration function or data from field measurements stored in a file on the initial time step, and then use the corrected values in the model itself, for example, with PresetAcceleration.

The corrections would be performed by subtracting a least squares polynomial from the nominal time histories, which results in a nearly equivalent vibration signal of the drifted one, but relative to the datum horizontal axis. Polynomial fits can optionally be made to all 3 kinematic variables or applied selectively. In addition, the order of the polynomial could be specified. This flexibility would allow for the user to achieve their desired result. The details for the theory and use of this vpp object would all be explained in the website's syntax page. I have created a prototype of this object on my [personal MOOSE Application](https://github.com/crswong888/scorpion/blob/master/src/vectorpostprocessors/LeastSquaresBaselineCorrection.C).

## Impact
Technically, this type of signal processing could be performed by a user them self, outside of MASTODON, and it has no direct relationship to finite element analysis. However, the numerical procedure is somewhat difficult and not generally known to structural engineers. Thus it would only serve to make MASTODON a more useful tool. It is also well-suited for automation. Many commercial software packages, such as [Seismosignal](https://seismosoft.com/products/seismosignal/) and [Abaqus](https://abaqus-docs.mit.edu/2017/English/SIMACAETHERefMap/simathe-c-baselinecorr.htm) have this capability.

@cbolisetti If you have the time and resources to review my proposed work, I wish to proceed with a PR.

### PR DESCRIPTION
**Pull request by crswong888**
@cbolisetti and @somu15,

I created a `Function` object, `BaselineCorrection`, to perform baseline correction on acceleration time histories. The object behaves like `PiecewiseLinear`, except the input is first adjusted according to the least squares method for baseline correction, and then linearly interpolated over time before outputting. A `BaselineCorrection` object can also be passed to a `PresetAcceleration` object to be used as a BC, so long as the initial acceleration is properly set by a `FunctionIC` object. If this PR is merged, I could come back and create an action system that handles all three of these objects.

I also created several TestHarness objects for `BaselineCorrection`, many of which are described in detail in the MooseDocs syntax page I wrote. The syntax page may seem usually long, mostly, because I want to be sure that MASTODON developers understand how I came up with this code, and what its use is. My approach is also somewhat of a novel one that builds upon some references which I mentioned in the docs. I will likely try to publish a paper that describes my method, at which point, I can come back and add a reference to it. I'll have to describe some verifications I've made by comparing against SeismoSignal and Abaqus, and so I will create an SQA page for `BaselineCorrection` at that point as well. 

I started a separate `dict()` for the MooseDocs Acronyms Extension at `doc/acronyms.yml` and updated the `config.yml` file to include both the one from MOOSE and this new one. This way, anyone can start adding acronyms that are unique to MASTODON documentation if they need to. This change was @aeslaughter approved :heavy_check_mark:

### Interesting Demos

I describe several demonstrations in the syntax page at `doc/content/source/functions/BaselineCorrection.md`, which I encourage you guys' to run a live serve of on your own machines, but a couple of them are particularly interesting:

1. The test at `test/tests/functions/baselinecorrection/twenty_sines_with_bc.i` applies a baseline corrected BC. I also ran this input file without baseline correction, and I rendered the two solutions, simultaneously, in the following GIF:

demo_5

2. I found a raw record of a Chi-Chi earthquake accelerogram from the [Strong Motion Database VDC](https://strongmotioncenter.org/) and I applied `BaselineCorrection` to it in the test at `test/tests/functions/baselinecorrection/chichi_A01263000_N.i`. I used Newmark integration to get the displacements from both the raw record and the corrected one and superimposed their time history plots:

newplot

### TODOs

The following items should be addressed in a future PR:

- Create action system for applying a baseline corrected acceleration BC
- Include reference to my paper describing this method when/if it becomes available
- Write a V&V report on this object which compares a MASTODON result to reputable third-parties, e.g., PEER, Abaqus, and/or SeismoSignal

(closes #296)

### CODE REVIEW

**Comment by crswong888** [comment]
I still need to create unit tests for the `BaselineCorrectionUtils`.

Also, since the syntax page is so long, perhaps we should discuss reorganizing it. Maybe I can move the part where I talk about the numerical implementation to the theory manual and move the part where I present demonstrations to examples.

File: include/functions/BaselineCorrection.h [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+  /// Newmark integration parameters
+  const Real & _gamma;
+  const Real & _beta;
+
+  /// order used for the least squares polynomial fit
+  unsigned int _order;
+
+  /// acceleration time history variables from input
+  std::vector<Real> _time;
+  std::vector<Real> _accel;
```
Comment (somu15) [inline include/functions/BaselineCorrection.h:50]: The input can be any time series, not just accelerations, correct? If so, you may consider changing the variable name from `_accel`

File: include/functions/BaselineCorrection.h [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+
+  /// adjusted (corrected) acceleration ordinates
+  std::vector<Real> _adj_accel;
```
Comment (somu15) [inline include/functions/BaselineCorrection.h:53]: Same comment as the previous one. If the input can be any time series, consider changing the name from `_adj_accel`

File: include/functions/BaselineCorrection.h (hunk 1 above)
Comment (crswong888) [reply include/functions/BaselineCorrection.h:50]: @somu15 Actually, no, the input may only be an acceleration time series. Baseline Correction isn't really necessary when the lower-order derivatives are the given.

File: include/functions/BaselineCorrection.h [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+
+  /// object to output linearly interpolated corrected acceleration ordinates
+  std::unique_ptr<LinearInterpolation> _linear_interp;
+
+  /// function value scale factor
```
Comment (somu15) [inline include/functions/BaselineCorrection.h:58]: Is this a scale factor multiplied to the input time series? If so, consider improving the description of this variable.

File: include/functions/BaselineCorrection.h [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+  const Real & _scale_factor;
+
+private:
+  /// Applies baseline correction to raw acceleration time history
```
Comment (somu15) [inline include/functions/BaselineCorrection.h:62]: Same comment as one of the previous ones. Re-consider the use of "acceleration time history."

File: include/functions/BaselineCorrection.h [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+  void applyCorrection();
+
+  /// Reads and builds data from supplied CSV file
+  void buildFromFile();
+
+  /// Builds data from pairs of `time_values` and `acceleration_values'
```
Comment (somu15) [inline include/functions/BaselineCorrection.h:68]: Same comment as one of the previous ones. Re-consider the use of "acceleration_values."

File: include/utils/BaselineCorrectionUtils.h [hunk 6]
```diff
@@ -0,0 +1,71 @@
...
+#include "libmesh/dense_vector.h"
+
+/**
+ * This namespace contains the functions used for the calculations corresponding
+ * to the time history adjustment procedure in BaselineCorrection
+ **/
+namespace BaselineCorrectionUtils
+{
+/// Evaluates an integral over a single time step with Newmark-beta method
+/// Also is used as simple trapezoidal rule when gamma = 0.5.
```
Comment (somu15) [inline include/utils/BaselineCorrectionUtils.h:31]: Consider changing to `Reduces to a simple trapezoidal rule when gamma = 0.5`

File: include/utils/BaselineCorrectionUtils.h [hunk 7]
```diff
@@ -0,0 +1,71 @@
...
+
+/// Evaluates a double integral over a single time step with Newmark-beta method
+Real newmarkBetaIntegrate(const Real & u_ddot_old,
+                          const Real & u_ddot,
+                          const Real & u_dot_old,
+                          const Real & u_old,
+                          const Real & beta,
+                          const Real & dt);
+
+/// Solves linear normal equation for minimum acceleration square error
```
Comment (somu15) [inline include/utils/BaselineCorrectionUtils.h:46]: Is there a simpler explanation to the phrase `linear normal equation`?

File: include/functions/BaselineCorrection.h (hunk 3 above)
Comment (crswong888) [reply include/functions/BaselineCorrection.h:58]: @somu15 the scale factor is applied to the output acceleration, i.e., it multiplies the baseline corrected values. I will be more verbose about this.

File: src/functions/BaselineCorrection.C [hunk 8]
```diff
@@ -0,0 +1,247 @@
...
+      "disp_fit_order",
+      "(0 <= disp_fit_order) & (disp_fit_order < 10)",
+      "If this is provided, the acceleration time history will be adjusted using an nth-order "
+      "polynomial fit of the nominal displacement data, where n = disp_fit_order (only integer "
+      "values from 0 to 9 are supported).");
+
+  params.addParam<Real>("scale_factor",
+                        1.0,
+                        "A scale factor to be applied to the adjusted acceleration time history.");
+  params.declareControllable("scale_factor");
```
Comment (somu15) [inline src/functions/BaselineCorrection.C:74]: Is this for the stochastic simulations stuff? If so, cool!

File: src/functions/BaselineCorrection.C [hunk 9]
```diff
@@ -0,0 +1,247 @@
...
+
+    for (unsigned int i = 0; i <= index_end; ++i)
+    {
+      p_fit = BaselineCorrectionUtils::computePolynomials(_order, coeffs, _time[i]);
+      _adj_accel[i] -= p_fit[0];
+      adj_disp[i] -= p_fit[2];
+    }
+  }
+
+  // adjust with displacement fit
```
Comment (somu15) [inline src/functions/BaselineCorrection.C:193]: Why are there three options for adjusting the fit? That is, accel, vel, and displacement. I am trying to understand the differences between these.

File: src/functions/BaselineCorrection.C [hunk 10]
```diff
@@ -0,0 +1,247 @@
...
+}
+
+void
+BaselineCorrection::buildFromFile()
+{
+  // Read data from CSV file
+  MooseUtils::DelimitedFileReader reader(getParam<FileName>("data_file"), &_communicator);
+  reader.read();
+
+  // Check if specific column headers were input
```
Comment (somu15) [inline src/functions/BaselineCorrection.C:214]: Do you need to always have the column header names? Currently, as far as I know, when we input accelerations to MASTODON, we are not required to have column header names.

File: test/tests/functions/baselinecorrection/chichi_A0126300_N.i [hunk 11]
```diff
@@ -0,0 +1,102 @@
+# A test of the BaselineCorrection function applied to an earthquake
```
Comment (somu15) [inline test/tests/functions/baselinecorrection/chichi_A0126300_N.i:1]: Having many tests is not necessarily bad. But, do you think you can consolidate the number of test without sacrificing much the comprehensiveness?

File: src/functions/BaselineCorrection.C (hunk 10 above)
Comment (crswong888) [reply src/functions/BaselineCorrection.C:214]: No this is not required, but, here, I enforced the case where if one variable was specified a header (e.g., time), then the other variable (acceleration), must also be specified a column header. This is to not confused the `DelimitedFileReader` class. If no header is specified for either, then the default is to read the first column as time and the second column as acceleration.

File: test/tests/functions/baselinecorrection/chichi_A0126300_N.i (hunk 11 above)
Comment (crswong888) [reply test/tests/functions/baselinecorrection/chichi_A0126300_N.i:1]: @somu15 yes

File: src/functions/BaselineCorrection.C (hunk 8 above)
Comment (crswong888) [reply src/functions/BaselineCorrection.C:74]: The scale factor is mostly for whatever purpose. Most MOOSE `Function` objects do this.

File: test/tests/functions/baselinecorrection/half_sine.i [hunk 12]
```diff
@@ -0,0 +1,40 @@
+# A simple test of the BaselineCorrection function output. The input acceleration
```
Comment (somu15) [inline test/tests/functions/baselinecorrection/half_sine.i:1]: Consider keeping either full sine or half sine test.

File: test/tests/functions/baselinecorrection/half_sine_scaled.i [hunk 13]
```diff
@@ -0,0 +1,41 @@
+# A test of the BaselineCorrection function output scaled by an input factor.
```
Comment (somu15) [inline test/tests/functions/baselinecorrection/half_sine_scaled.i:1]: Or maybe just include half sine scaled and consider removing half sine and full sine.

File: test/tests/functions/baselinecorrection/chichi_A0126300_N.i (hunk 11 above)
Comment (crswong888) [reply test/tests/functions/baselinecorrection/chichi_A0126300_N.i:1]: @somu15 consider taking a moment to read my MooseDocs page. The tests each demonstrate a different aspect of this code. But, maybe there's someway I can consolidate all of the `mooseError()` tests into a single input file.

File: test/tests/functions/baselinecorrection/twenty_sines.i [hunk 14]
```diff
@@ -0,0 +1,43 @@
+# A comprehensive test of the BaselineCorrection function output. The input
```
Comment (somu15) [inline test/tests/functions/baselinecorrection/twenty_sines.i:1]: Or maybe just keep twenty sines scaled and remove half sine scaled, half sine, and full sine.

File: test/tests/functions/baselinecorrection/chichi_QT063000_V.i [hunk 15]
```diff
@@ -0,0 +1,102 @@
+# A test of the BaselineCorrection function applied to an earthquake
```
Comment (somu15) [inline test/tests/functions/baselinecorrection/chichi_QT063000_V.i:1]: Consider including only either one of the earthquake input tests.

File: test/tests/functions/baselinecorrection/half_sine_scaled.i (hunk 13 above)
Comment (crswong888) [reply test/tests/functions/baselinecorrection/half_sine_scaled.i:1]: @somu15 what if I put all of these tests into a single input file? They could all be operated on as separate function objects, and be written to separate post-processors.

File: include/utils/BaselineCorrectionUtils.h (hunk 6 above)
Comment (crswong888) [reply include/utils/BaselineCorrectionUtils.h:31]: @somu okay

File: include/utils/BaselineCorrectionUtils.h (hunk 7 above)
Comment (crswong888) [reply include/utils/BaselineCorrectionUtils.h:46]: @somu15 This is what it is referred to as in terms of linear algebra jargon. For example, see [section 8.2 (page 510) of this textbook](https://fac.ksu.edu.sa/sites/default/files/numerical_analysis_9th.pdf), or see this [Wikipedia page](https://en.wikipedia.org/wiki/Polynomial_least_squares). 

It's just the equation that solves the least squares polynomial. I suppose I could just straight-up call it this instead.

File: include/utils/BaselineCorrectionUtils.h (hunk 6 above)
Comment (crswong888) [reply include/utils/BaselineCorrectionUtils.h:31]: @somu15 okay

File: doc/content/source/functions/BaselineCorrection.md [hunk 16]
```diff
@@ -0,0 +1,409 @@
+# Baseline Correction
+
+`BaselineCorrection` is a [`Function`](syntax/Functions/index.md) object that applies a [!ac](BLC) to an acceleration time history, $\ddot{u}(t)$, and outputs the adjusted values. The corrections are performed by first computing the nominal (*uncorrected*) velocity, $\dot{u}$, and displacement, $u$, by the [Newmark-beta method](manuals/theory/index.md#time-integration). An $n$-th order polynomial approximation of the acceleration, velocity, and/or displacement is then found by the method of least squares and each are successively subtracted from the nominal acceleration to obtain the corrected one, $\ddot{u}^{\ast}$. That is, the corrected acceleration time history is given by the following:
```
Comment (somu15) [inline doc/content/source/functions/BaselineCorrection.md:3]: `An $n$-th order polynomial approximation of the acceleration, velocity, and/or displacement is then found by the method of least squares and each are successively subtracted from the nominal acceleration to obtain the corrected one, $\ddot{u}^{\ast}$.` looks a bit confusing and you may want to break it up. Also, consider explaining more `each are subtracted from`.

File: src/functions/BaselineCorrection.C (hunk 9 above)
Comment (crswong888) [reply src/functions/BaselineCorrection.C:193]: @somu15 this I cannot justify in a GitHub comment, I would need to give you a phone call.

File: doc/content/source/functions/BaselineCorrection.md (hunk 16 above)
Comment (crswong888) [reply doc/content/source/functions/BaselineCorrection.md:3]: @somu15 the next section about the numerical implementation explains this. Please read on.

File: doc/content/source/functions/BaselineCorrection.md [hunk 17]
```diff
@@ -0,0 +1,409 @@
...
+
+The order of the approximating polynomials, $n$, which are set by the `accel_fit_order`, `vel_fit_order`, or `disp_fit_order` parameters, can be assigned a maximum integer value of nine, since the computations typically become unstable when attempting to find polynomials of orders greater than nine. The corrections are highly sensitive to the approximation order, and so it is recommended that a user run with several trial values until the desired result is achieved. In general, the order should be less than that which is required to produce a true best-fit curve of $\ddot{u}(t)$. For example, in [example-2], a half-sine wave was input. Since a half-sine wave could be, nearly, perfectly approximated by a certain quadratic function, the order for the corrections was set to one. However, for higher order inputs, such as the 20-cycle sine wave used in [example-2], finding the right correction configuration becomes tricky, and is a matter of user-preference. The [#demo] section provides several examples that should clarify why it is important to choose a suitable polynomial order.
+
+!alert tip title=Use the Fewest Polynomials with the Lowest Orders Possible
```
Comment (somu15) [inline doc/content/source/functions/BaselineCorrection.md:161]: Can you please describe what happens if a user does `overcorrection`?

Comment (crswong888) [reply doc/content/source/functions/BaselineCorrection.md:161]: @somu15 yes

File: doc/content/source/functions/BaselineCorrection.md [hunk 18]
```diff
@@ -0,0 +1,409 @@
...
+
+!plot scatter filename=baselinecorrection_data/demo_4_adjusted.csv
+              data=[{'x':'time', 'y':'displacement', 'name':'Corrected Displacement'}]
+              layout={'xaxis':{'title':'Time'},
+                      'yaxis':{'title':'Displacement'},
+                      'title':'Corrected Displacement Time History'}
+              id=demo-4-adjusted
+              caption=Plot of the displacement time history resulting from integration of the baseline corrected values of the acceleration shown in [demo-4].
+
+### Apply a Corrected Twenty-sine as a Boundary Condition id=demo-5
```
Comment (somu15) [inline doc/content/source/functions/BaselineCorrection.md:307]: If you consolidate `halfsine, fullsine, halfsinewithBC, twentysines, twentysineswithBC` to one, then consider modifying the this markdown file as well.

File: src/functions/BaselineCorrection.C (hunk 9 above)
Comment (crswong888) [reply src/functions/BaselineCorrection.C:193]: @somu15 this approach is explained in a paper: [https://journals.sagepub.com/doi/pdf/10.1177/1687814016675534](https://journals.sagepub.com/doi/pdf/10.1177/1687814016675534)

I included this as a reference on the MooseDocs page.

**Comment by somu15** [comment]
@crswong888 Nice PR. I made a few comments for now. Please make changes if you agree with them. @cbolisetti may have additional comments.

File: doc/content/source/functions/BaselineCorrection.md (hunk 18 above)
Comment (crswong888) [reply doc/content/source/functions/BaselineCorrection.md:307]: @somu15 Ill just reference the correct block in all of the `!listing` calls to the single, comprehensive input file. 

But, I should keep the input file that uses `PresetAcceleration` separate I think.

**Comment by crswong888** [comment]
@somu15 Thank you so much for taking the time to go through my work. I replied to some of your comments, perhaps you will continue some discussions there. I am in agreement with most of your comments and will make those changes.

**Comment by somu15** [comment]
Hi @crswong888 What's the status of this?

**Comment by crswong888** [comment]
@somu15 I still haven't gotten around to addressing your comments. I can do this by the end of next week, the latest, if this is okay.

**Comment by crswong888** [comment]
@somu15 I have addressed some of the things we have talked about. I have made some pretty significant changes, mainly, I've made it possible for the function to output the corrected values for any of the three types of time series: acceleration, velocity, or displacement, since there's no reason why this wasn't allowed. 

Please review my work when you get the chance.

**Comment by somu15** [comment]
Could you please exclude the .gif file? We thought that it's big, memory wise.

**Comment by somu15** [comment]
How much time are the baseline correction tests taking now, Chris?

File: test/tests/functions/baselinecorrection/tests [hunk 19]
```diff
@@ -0,0 +1,77 @@
+[Tests]
+  issues = '#296'
+  design = functions/BaselineCorrection.md
+
+  [test_correction]
+    requirement = "The BaselineCorrection class shall accurately adjust a raw acceleration time history according to the method of using least squares polynomial fits."
```
Comment (somu15) [inline test/tests/functions/baselinecorrection/tests:6]: The BaselineCorrection class shall accurately adjust a raw acceleration time history according to the method of using least squares polynomial fits for

File: test/tests/functions/baselinecorrection/tests [hunk 20]
```diff
@@ -0,0 +1,77 @@
...
+
+    [half_sine_scaled]
+      type = CSVDiff
+      input = half_sine_scaled.i
+      csvdiff = half_sine_scaled_out.csv
+
+      detail = "The output should be a displacement time history scaled by a factor of 4.886 that does not exhibit drifting behavior. The input acceleration time history is a half cycle of `9.87 \sin(\pi t)`."
```
Comment (somu15) [inline test/tests/functions/baselinecorrection/tests:13]: a half-sine acceleration input.

File: test/tests/functions/baselinecorrection/tests [hunk 21]
```diff
@@ -0,0 +1,77 @@
...
+    []
+    [twenty_sines_with_bc]
+      type = CSVDiff
+      input = twenty_sines_with_bc.i
+      csvdiff = twenty_sines_with_bc_out.csv
+
+      detail = "The output should be an adjusted acceleration time history that is compatible with ICs and BCs objects and the displacement at the mesh boundary where it is prescribed should not exhibit drifting behavior. The input acceleration time history is twenty cycles of `-250 \pi^{2} \sin(50 \pi t)`."
```
Comment (somu15) [inline test/tests/functions/baselinecorrection/tests:20]: for a twenty sines acceleration input.

File: test/tests/functions/baselinecorrection/tests [hunk 22]
```diff
@@ -0,0 +1,77 @@
...
+    []
+  []
+
+  [test_errors]
+    requirement = "The BaselineCorrection class shall appropriately invoke `mooseError()` when the input configuration is not supported."
```
Comment (somu15) [inline test/tests/functions/baselinecorrection/tests:25]: The BaselineCorrection class shall appropriately invoke `mooseError()` when

File: test/tests/functions/baselinecorrection/tests [hunk 23]
```diff
@@ -0,0 +1,77 @@
...
+
+    [no_input]
+      type = RunException
+      input = no_input_error.i
+      expect_err = "In BaselineCorrection corrected_accel_func: Either `data_file` or `time_values` and `acceleration_values` must be specified exclusively."
+
+      detail = "This error shall occur since no acceleration time history was provided."
```
Comment (somu15) [inline test/tests/functions/baselinecorrection/tests:32]: no acceleration time history is provided.

File: test/tests/functions/baselinecorrection/tests [hunk 24]
```diff
@@ -0,0 +1,77 @@
...
+    []
+    [extra_input]
+      type = RunException
+      input = extra_input_error.i
+      expect_err = "In BaselineCorrection corrected_accel_func: Either `data_file` or `time_values` and `acceleration_values` must be specified exclusively."
+
+      detail = "This error shall occur since the parameters which control the different input types were not provided exclusively."
```
Comment (somu15) [inline test/tests/functions/baselinecorrection/tests:39]: the parameters which control the different input types are not provided.

File: test/tests/functions/baselinecorrection/tests [hunk 25]
```diff
@@ -0,0 +1,77 @@
...
+    []
+    [missing_value_param]
+      type = RunException
+      input = missing_value_param_error.i
+      expect_err = "In BaselineCorrection corrected_accel_func: Either `data_file` or `time_values` and `acceleration_values` must be specified exclusively."
+
+      detail = "This error shall occur since the `time_values` and `acceleration_values` parameters were not provided together."
```
Comment (somu15) [inline test/tests/functions/baselinecorrection/tests:46]: the `time_values` and `acceleration_values` parameters are not provided together.

File: test/tests/functions/baselinecorrection/tests [hunk 26]
```diff
@@ -0,0 +1,77 @@
...
+    []
+    [missing_file_param]
+      type = RunException
+      input = missing_file_param_error.i
+      expect_err = "In BaselineCorrection corrected_accel_func: A column header name was specified for the for the time data. Please specify a header for the acceleration data using the 'acceleration_name' parameter."
+
+      detail = "This error shall occur since the `time_name` and `acceleration_name` parameters were not provided together."
```
Comment (somu15) [inline test/tests/functions/baselinecorrection/tests:53]: the `time_name` and `acceleration_name` parameters are not provided together.

File: test/tests/functions/baselinecorrection/tests [hunk 27]
```diff
@@ -0,0 +1,77 @@
...
+    []
+    [size_mismatch]
+      type = RunException
+      input = size_mismatch_error.i
+      expect_err = "In BaselineCorrection corrected_accel_func: The length of time and acceleration data must be equal."
+
+      detail = "This error shall occur since the lengths of the time data and acceleration data are not equal"
```
Comment (somu15) [inline test/tests/functions/baselinecorrection/tests:60]: the lengths of the time data and acceleration data are not equal.

File: test/tests/functions/baselinecorrection/tests [hunk 28]
```diff
@@ -0,0 +1,77 @@
...
+    []
+    [no_data]
+      type = RunException
+      input = no_data_error.i
+      expect_err = "In BaselineCorrection corrected_accel_func: The length of time and acceleration data must be > 0."
+
+      detail = "This error shall occur since no data was provided at all."
```
Comment (somu15) [inline test/tests/functions/baselinecorrection/tests:67]: no data is provided at all.

File: test/tests/functions/baselinecorrection/tests [hunk 29]
```diff
@@ -0,0 +1,77 @@
...
+    []
+    [no_order]
+      type = RunException
+      input = no_order_error.i
+      expect_err = "In BaselineCorrection corrected_accel_func: No values were input for parameters 'accel_fit_order', 'vel_fit_order', or 'disp_fit_order'. Please specify an integer value from 0 to 9 for at least one of these parameters."
+
+      detail = "This error shall occur since no least squares polynomials would be computed and, therefore, no adjustments would be made."
```
Comment (somu15) [inline test/tests/functions/baselinecorrection/tests:74]: no least squares polynomials can be computed and, therefore, no adjustments can be made.

**Review by somu15 — COMMENTED** [review]
@cbolisetti I am done with my review. I have some additional comments for Chris.

File: src/functions/BaselineCorrection.C [hunk 30]
```diff
@@ -0,0 +1,267 @@
...
+  // size checks
+  if (_time.size() != _accel.size())
+    mooseError("In BaselineCorrection ",
+               _name,
+               ": The length of time and acceleration data must be equal.");
+  if (_time.size() == 0 || _accel.size() == 0)
+    mooseError(
+        "In BaselineCorrection ", _name, ": The length of time and acceleration data must be > 0.");
+
+  // check that at lease one least squares fit will be applied
```
Comment (cbolisetti) [inline src/functions/BaselineCorrection.C:115]: *at least

**Review by cbolisetti — CHANGES_REQUESTED** [review]

**Comment by crswong888** [comment]
@cbolisetti @somu15 i will be addressing your comments very soon

**Comment by crswong888** [comment]
@cbolisetti @somu15 I am going to close this PR and start from scratch. My algorithm has changed significantly.

**Pull request closed without merging**